File: recipes/subms-events-store/python/src/subms_events_store/recipe.py
"""Perf runner emitting the SubMsBenchSummary JSON contract (lang=python).
Stages: append / replay / catch_up. Run:

    PYTHONPATH="src;../../subms-events/python/src" python -m subms_events_store.recipe > ../perf/python.json
"""

from __future__ import annotations

import json
import sys
import time

from subms_events import Event

from .event_store import EventStore
from .projection import Projector, replay

REPLAY_N = 1_000


def _summarize(samples):
    s = sorted(samples)
    n = len(s)

    def pct(p):
        return s[min(n - 1, int(p * n))]

    step = max(1, n // 500)
    return {
        "count": n,
        "p50_ns": pct(0.50),
        "p99_ns": pct(0.99),
        "p999_ns": pct(0.999),
        "max_ns": s[-1],
        "mean_ns": int(sum(s) / n),
        "samples_ns": s[::step][:500],
    }


def run(entries: int = 20_000) -> dict:
    store = EventStore()
    append = []
    for i in range(entries):
        ev = Event.builder("evt").at("t").attr("i", str(i)).build()
        t0 = time.perf_counter_ns()
        store.append(ev)
        append.append(time.perf_counter_ns() - t0)

    base = EventStore()
    for _ in range(REPLAY_N):
        base.append(Event.builder("evt").at("t").build())
    rep = []
    for _ in range(entries):
        t0 = time.perf_counter_ns()
        replay(base, 0, lambda n, e: n + 1)
        rep.append(time.perf_counter_ns() - t0)

    proj = Projector(0)
    proj.catch_up(base, lambda n, e: n + 1)
    cu = []
    for _ in range(entries):
        base.append(Event.builder("x").at("t").build())
        t0 = time.perf_counter_ns()
        proj.catch_up(base, lambda n, e: n + 1)
        cu.append(time.perf_counter_ns() - t0)

    return {
        "workload": "subms-events-store",
        "lang": "python",
        "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "inputs": {"entries": str(entries)},
        "meta": {
            "subms.recipe.slug": "subms-events-store",
            "subms.recipe.category": "storage",
            "replay_window": str(REPLAY_N),
            "subms.workload.feature": "in-memory-event-sourcing",
        },
        "stages": {
            "append": _summarize(append),
            "replay": _summarize(rep),
            "catch_up": _summarize(cu),
        },
    }


def main() -> None:
    entries = int(sys.argv[1]) if len(sys.argv) > 1 else 20_000
    json.dump(run(entries), sys.stdout, separators=(",", ":"))


if __name__ == "__main__":
    main()

File: recipes/subms-events-store/rust/src/recipe.rs
//! `SubMsRecipe` impl. Stages: `append` (push + notify), `replay` (full fold over
//! a fixed log), `catch_up` (incremental projection of the tail).

use subms::{SubMsBenchParams, SubMsPerfHarness, SubMsRecipe, SubMsStageKind, SubMsTimer};

use crate::{Event, EventStore, Projector, replay};

pub struct EventStoreRecipe;

const REPLAY_N: usize = 1_000;

impl SubMsRecipe for EventStoreRecipe {
    fn name(&self) -> &str {
        "subms-events-store"
    }

    fn run(&self, h: &mut SubMsPerfHarness, params: &SubMsBenchParams) {
        let entries = params.entries;

        // append: into a fresh store.
        let mut store = EventStore::new();
        let s_app = h
            .stage("append", entries)
            .with_kind(SubMsStageKind::HotPath);
        for i in 0..entries {
            let ev = Event::builder("evt")
                .at("t")
                .attr("i", &i.to_string())
                .build();
            let t0 = SubMsTimer::tick();
            store.append(ev);
            s_app.record(t0.elapsed_ns());
        }

        // replay: full fold over a fixed REPLAY_N-event store.
        let mut base = EventStore::new();
        for _ in 0..REPLAY_N {
            base.append(Event::builder("evt").at("t").build());
        }
        let s_rep = h
            .stage("replay", entries)
            .with_kind(SubMsStageKind::BatchOp);
        for _ in 0..entries {
            let t0 = SubMsTimer::tick();
            let count = replay(&base, 0u64, |n, _e| *n += 1);
            s_rep.record(t0.elapsed_ns());
            std::hint::black_box(count);
        }

        // catch_up: incremental projection of one new event each round.
        let mut proj = Projector::new(0u64);
        proj.catch_up(&base, |n, _e| *n += 1);
        let s_cu = h
            .stage("catch_up", entries)
            .with_kind(SubMsStageKind::HotPath);
        for _ in 0..entries {
            base.append(Event::builder("x").at("t").build());
            let t0 = SubMsTimer::tick();
            proj.catch_up(&base, |n, _e| *n += 1);
            s_cu.record(t0.elapsed_ns());
        }

        h.add_meta("replay_window", &REPLAY_N.to_string());
        h.add_meta("final_log_len", &base.len().to_string());
        h.add_meta("subms.workload.feature", "in-memory-event-sourcing");
    }
}
